File: day06/guard.py
#!/usr/bin/env python3
import sys

def simulate_guard_path(grid):
    directions = {'^': (-1, 0), '>': (0, 1), 'v': (1, 0), '<': (0, -1)}
    turn_right = {'^': '>',
                  '>': 'v',
                  'v': '<',
                  '<': '^'}
    
    # Find the guard's starting position and orientation
    rows, cols = len(grid), len(grid[0])
    guard_pos = None
    guard_dir = None
    
    for r in range(grid):
        assert len(grid[r]) == cols
        for c in range(cols):
            if grid[r][c] in directions:
                guard_pos = (r, c)
                guard_dir = grid[r][c]
                break
        if guard_pos:
            break
    
    # Track visited positions and simulate the guard's movement
    visited = set()
    visited.add(guard_pos)
    
    while True:
        # Calculate the next position
        dr, dc = directions[guard_dir]
        next_pos = (guard_pos[0] + dr, guard_pos[1] + dc)
        
        # Check if the next position is outside the grid
        if not (0 <= next_pos[0] < rows and 0 <= next_pos[1] < cols):
            break
        
        # Check if the next position is an obstacle
        if grid[next_pos[0]][next_pos[1]] == '#':
            # Turn right if there's an obstacle
            guard_dir = turn_right[guard_dir]
        else:
            # Move forward if the next position is free
            guard_pos = next_pos
            visited.add(guard_pos)
    
    return len(visited)

# Read input from stdin
if __name__ == "__main__":
    grid = [list(line.strip()) for line in sys.stdin]
    result = simulate_guard_path(grid)
    print(f"The guard visits {result} distinct positions.")

File: day06/guard.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE QuasiQuotes #-}

import AoC
import Control.Applicative
import Control.Exception
import Control.Monad
import Control.Monad.ST
-- import qualified Data.Algebra.Boolean as B
import Data.Char
import Data.List
import Data.Maybe
import Data.Function
import Data.Array
import Data.Array.ST
import Data.Tuple.Extra
import Data.Graph
import Data.Set (Set)
import qualified Data.Set as Set

-- import GHC.Utils.Misc

import Debug.Trace
import qualified Data.Text as T
import qualified Data.Sequence as S
import Text.ParserCombinators.ReadP hiding (count)
import Text.Printf
-- import Text.Scanf



-- If part1 and part2 are very different,
-- toggle which part to compute with this flag.
part2 = False


-- Check if the position is an obstacle ('#')
isObstacle :: Coord -> CharMap -> Bool
isObstacle pos grid = grid!pos == '#'

-- Simulate the guard's path and collect visited positions
simulateGuard :: CharMap -> Coord -> Cardir -> Set Coord -> Set Coord
simulateGuard grid pos dir visited
  | not (inBounds grid nextPos) = visited -- Guard moves out of bounds
  | isObstacle nextPos grid       = simulateGuard grid pos (turnRight dir) visited -- Obstacle encountered, turn right
  | otherwise                     = simulateGuard grid nextPos dir (Set.insert nextPos visited) -- Move forward
  where
    nextPos = move pos dir

-- Parse the grid to find the guard's initial position and direction
findGuard :: CharMap -> Maybe (Coord, Cardir)
findGuard grid = foldl' findFold Nothing $ assocs grid
  where findFold (Just d) (_, _) = Just d
        findFold Nothing (_, '.') = Nothing
        findFold Nothing (_, '#') = Nothing
        findFold Nothing (pos, c) = Just $ (pos, chDir c)

-- Main function to calculate the number of distinct positions visited
countVisitedCoords :: CharMap -> Int
countVisitedCoords grid =
  let ((startX, startY), dir) = fromJust $ findGuard grid
      visited = simulateGuard grid (startX, startY) dir (Set.singleton (startX, startY))
  in Set.size visited

-- Example usage
main :: IO ()
main = do
  -- let grid = [ "....#.....",
               -- ".........#",
               -- "..........",
               -- "..#.......",
               -- ".......#..",
               -- "..........",
               -- ".#..^.....",
               -- "........#.",
               -- "#.........",
               -- "......#..." ]
  grid <- getCharMap
 
  print $ countVisitedCoords grid
